source/common/network/udp_listener_impl.cc

```diff
@@ -94,56 +71,75 @@ void UdpListenerImpl::onSocketEvent(short flags) {
 
 void UdpListenerImpl::handleReadCallback() {
   ENVOY_UDP_LOG(trace, "handleReadCallback");
-  sockaddr_storage addr;
-  socklen_t addr_len = 0;
-
+  // TODO(danzh) make this variable configurable to support jumbo frames.
+  const uint64_t read_buffer_length = MAX_UDP_PACKET_SIZE;
   do {
-    ReceiveResult recv_result = doRecvFrom(addr, addr_len);
-    if ((recv_result.result_.rc_ < 0)) {
-      if (recv_result.result_.errno_ != EAGAIN) {
-        ENVOY_UDP_LOG(error, "recvfrom result {}", recv_result.result_.errno_);
+    Buffer::InstancePtr buffer = std::make_unique<Buffer::OwnedImpl>();
+    Buffer::RawSlice slice;
+    const uint64_t num_slices = buffer->reserve(read_buffer_length, &slice, 1);
+    ASSERT(num_slices == 1);
+
+    IoHandle::RecvMsgOutput output(&packets_dropped_);
+    uint32_t old_packets_dropped = packets_dropped_;
+    MonotonicTime receive_time = time_source_.monotonicTime();
+    Api::IoCallUint64Result result = socket_.ioHandle().recvmsg(
+        &slice, num_slices, socket_.localAddress()->ip()->port(), output);
+
+    if (!result.ok()) {
+      // No more to read or encountered a system error.
+      if (result.err_->getErrorCode() != Api::IoError::IoErrorCode::Again) {
+        ENVOY_UDP_LOG(error, "recvfrom result {}: {}",
+                      static_cast<int>(result.err_->getErrorCode()),
+                      result.err_->getErrorDetails());
         cb_.onReceiveError(UdpListenerCallbacks::ErrorCode::SyscallError,
-                           recv_result.result_.errno_);
+                           result.err_->getErrorCode());
       }
+      // Stop reading.
       return;
     }
 
-    if (recv_result.result_.rc_ == 0) {
+    if (result.rc_ == 0) {
       // TODO(conqerAtapple): Is zero length packet interesting?
-      return;
+      ENVOY_UDP_LOG(warn, "received 0-length packet");
```

Comment:
When does this happen? Anything in the data path needs to be debug or lower to avoid log spam. If we are concerned about this we should have a stat so I would add a TODO

Reply:
Good point, it happens when we receive empty udp packets, thus can cause log spam. I added a TODO to use stats.